### ISSUE TITLE
Call of function with side-effect is incorrectly identified as constant logical expression

### ISSUE THREAD

**Issue by jan-dubsky** [mgechev/revive#657]
**Describe the bug**

Multiple calls of a function with side effect are incorrectly evaluated as constant and linted as redundant.

I have a following example code:
```go
func main() {
	file, err := os.Open(os.Args[1])
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	if scanner.Scan() && scanner.Scan() {
		fmt.Printf("Second line of file is: %s\n", scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}
}
```
which is supposed to read the second line in the file.

This code is perfectly valid and the result of `scanner.Scan()` is definitely not constant as the call has effect in change of both the internal state of the scanner and the `io.Reader` (`file` in this case). 

**To Reproduce**
1. I updated revive `go get -u github.com/mgechev/revive`
2. I run it with the following flags & configuration file:

```shell
$ ~/go/bin/revive -config c.toml ./...
```

```toml
errorCode = 1

[rule.constant-logical-expr]
```

**Expected behavior**
The linter should see that duplicate call of a function with side effect is not a constant expression and it shouldn't report any lint error.

**Desktop (please complete the following information):**
 - OS: Arch Linux
 - go1.18

**Comment by chavacava** [comment]
Hi @jan-dubsky, thanks for filling the issue.
Currently, `revive` is unable to know if a function has a side effect or not (it is a hard problem in the general case). 
We could lower the confidence of the failures in the case of Boolean expressions including function calls. 
For the moment you can add a [comment directive](https://github.com/mgechev/revive#comment-directives) to disable the rule for the specific cases (I hope, there are not too many)

**Comment by jan-dubsky** [comment]
Hi @chavacava, thanks for the recommendation about disabling the linter. I use it as a quick hack solution.

What I wanted to propose is to make the linter conservative in what is considered an error. In this case, the expected behaviour for me would be that the linter won't considered function call a constant value. Possible future improvements could then include recognition of getters, pure functions etc., which would make the linter more accurate. But in general, I'd appreciate a philosophy that it's better for the linter to miss an problematic code than to report false-positive error. Mostly because false-positive errors decrease developer trust in linter errors in general, not mentioning the time developers have to invest in disabling the linter for those false-positive rows.

**Comment by chavacava** [comment]
I totally agree on the need to minimize the number of false positives.
It is something we control before adding new rules to `revive`.
In this particular case, I've re-run the rule against some large repos (kubernetes, docker, terraform, ...) and I got zero false positives (discarding the `x != x` used to check if a float is NaN)

About considering functions to be non-pure by default is something we could consider. I think that in this particular case (i.e. functions being called in the context of a logical expression) assuming functions to be pure is not a hard assumption. Why? Because, I think, calling non-pure functions in logical expressions is more the exception than the norm. From my experience, cases like the one in your (artificial?) example are not frequent enough to justify skipping the analysis of logical expressions with function calls. (I might be wrong, I will try to get some empirical data by analyzing the use of functions in logical expressions)

**Comment by chavacava** [comment]
Closing without fixing. These kind of false positive are likely to be very rare.
Will re-open if false positives are too noisy.
